pkg/build/tar.go

```diff
@@ -8,36 +8,54 @@ import (
 	"time"
 )
 
-func StreamLayer(imagePath string) (tarReader io.ReadCloser, errChan chan error) {
-	errChan = make(chan error, 1)
-	reader, writer := io.Pipe()
-	tarWriter := tar.NewWriter(writer)
-
-	go func() {
-		defer writer.Close()
-		defer tarWriter.Close()
-		err := addFileToTarWriter(imagePath, tarWriter)
-		if err != nil {
-			errChan <- fmt.Errorf("error adding file '%s', to tarball: %v", imagePath, err)
-		}
-		close(errChan)
-	}()
+func StreamLayerOpener(imagePath string) func() (io.ReadCloser, error) {
+	modTime := time.Now()
 
-	return reader, errChan
-}
+	return func() (io.ReadCloser, error) {
+		fileErrorChan := make(chan error)
+		pipeReader, pipeWriter := io.Pipe()
 
-func addFileToTarWriter(filePath string, tarWriter *tar.Writer) error {
-	file, err := os.Open(filePath)
-	if err != nil {
-		return fmt.Errorf("error opening file: %v", err)
-	}
-	defer file.Close()
+		go func() {
+			defer pipeWriter.Close()
 
-	stat, err := file.Stat()
-	if err != nil {
-		return fmt.Errorf("error getting file information with stat: %v", err)
+			file, err := os.Open(imagePath)
+			if err != nil {
+				fileErrorChan <- fmt.Errorf("error opening file: %w", err)
+				return
+			}
+			defer file.Close()
+
+			stat, err := file.Stat()
+			if err != nil {
+				fileErrorChan <- fmt.Errorf("error getting file information with stat: %w", err)
+				return
+			}
+
+			// Close channel after successfully opening file to avoid deadlock
+			close(fileErrorChan)
+
+			tarWriter := tar.NewWriter(pipeWriter)
+			err = addFileToTarWriter(file, stat, modTime, tarWriter)
+			if err != nil {
+				// Move the error to the PipeReader side. It is ok to call close on PipeWriter multiple times.
+				pipeWriter.CloseWithError(fmt.Errorf("error adding file '%s', to tarball: %w", imagePath, err))
+			}
+			err = tarWriter.Close()
```

Comment:
The `tarWriter` should be closed before the `pipeWriter`, because it writes footer on close.

Reply:
If err is nil then `tarWriter` is closed before `pipeWriter` (it is closed by line 19).

If an error happens in`addFileToTarWriter`, there is probably no point in keeping the order?

Reply:
ok